_posts/2025-04-25-from-recipes-to-vectors-building-a-hybrid-search-app-with-opensearch.md

```diff
@@ -0,0 +1,299 @@
+```
+
+### Semantic Search
+
+However, if I wanted to find chicken recipes that were asian inspired and featured noodles and peanuts, a simple keyword search wouldn’t be enough. Keyword-based retrieval relies on exact matches, meaning it would only return documents that explicitly contain those words—missing relevant results that describe the same concept in different terms.
+
+To retrieve the most relevant recipes based on meaning rather than exact wording, we need to use semantic search.
+
+First of all, we would need to pass our search query through the same embedding model (e.g., Cohere's `embed-english-v3`), so we can use the resulting vector as the vector in an OpenSearch knn query.
+
+Once we have the query embedding, we can use it in a **knn query**, which in this case, an additional type of a query OpenSearch supports, using the knn plugin. 
+
+```
+POST /recipes/_search
+{
+   "size": 5,
+   "query": {
+       "knn": {
+           "embedding": {
+               "vector": question_embedding,
+               "k": 5
+           }
+       }
+   }
+}
+
+-----
+Curried Peanut Chicken Nibbles
+Spicy Chicken Salad
+Kung Pao Chicken
+Linguine with Chicken and Peanut Sauce
+Spicy Thai Noodles
+-----
+```
+
+Breaking down the parameters of the knn query:
+
+* `field: "embedding"` - The vector field where embeddings are stored.
+* `vector: [...]` - The embedding of our search query, generated using the same model as our dataset.
+* `k: 5` - Retrieve the 5 most similar recipes based on vector similarity.
+
+OpenSearch also allows you to omit k if you specify either `max_distance` or `min_score` as alternative ways to control the amount of documents to search semantically.
+
+This approach allows OpenSearch to find recipes that may not contain the exact words "noodles and peanuts" explicitly but definitely fit into the Asian category — such as "Kung Pao Chiken”.
+
+### Hybrid search
+
+
+Lastly, let’s say we wanted to cook a low sugar dessert. If I searched for this semantically, I might get good results, but some other low sugar dishes might creep in (*Is sweet potato soup a dessert?*). In this case I could use hybrid search and combine the two types of queries together.
+
+Here’s an example of how to structure this query:
+
+```
+POST /recipes/_search 
+{
+ "size": 5,
+ "query": {
+   "bool": {
+     "must": [
+       {
+         "knn": {
+           "embedding": {
+             "vector": question_embedding,
+             "k": 5
+           }
+         }
+       }
+     ],
+     "filter": [
+       { "term": { "RecipeCategory": "Dessert" } },
+     ]
+   }
+ }
+}
+
+------
+No Fat Fruitcake
+Vanilla Pudding
+Low Fat Carrot Cake
+Summer Dessert
+Apple Pudding
+------
+
+```
+
+In this case OpenSearch filters out only Desserts and then looks for the most relevant results **semantically**. We could also use a should query to boost desserts in the search results but not exclude them entirely (*because maybe sweet potato soup IS a dessert after all?*)
```

Comment:
```suggestion
In this case OpenSearch filters out only desserts and then looks for the most relevant results **semantically**. We could also use a should query to boost desserts in the search results but not exclude them entirely (*because maybe sweet potato soup IS a dessert after all?*).
```